### PR TITLE
Feat: use btp plugin

### PR DESCRIPTION
**Pull request by sharafian**
WIP, makes ILP-plugin-xrp-paychan natively ILPv4 by basing it off of `ilp-plugin-btp` instead of `ilp-plugin-payment-channel-framework`. semantics around payment channels stay the same.

### CODE REVIEW

**Comment by michielbdejong** [comment]
Looks like in https://github.com/ripple/ilp-plugin-xrp-paychan/blob/bs-use-btp-plugin/test/scripts/test.js#L45 the server will wait until the client connects, but the client doesn't start connecting yet until the server is connected, so they end up waiting for each other. Can be fixed by putting them together in a `Promise.all`

**Comment by dappelt** [comment]
@sharafian What's the plan for merging this?

**Comment by sharafian** [comment]
It's currently being used by our connectors, but we should try to merge it soon. The main issue right now is that the tests don't fit into this new plugin so we'll need to rewrite them.

**Comment by dappelt** [comment]
Do you want to have a stab at the unit tests? I can assist with the integration tests. Let me know.

**Comment by dappelt-public** [comment]
@sharafian What's the status on this? It would be great if you get this merged so that I can patch some of the vulnerabilities we identified in asym-server.

**Comment by sharafian** [comment]
I'm working on the unit tests today. Should have it ready for review by tomorrow

File: index.js [hunk 1]
```diff
@@ -1,500 +1,411 @@
...
+  constructor (opts) {
+    super(opts)
+
+    this._xrpServer = opts.xrpServer || opts.rippledServer // TODO: deprecate rippledServer
+    this._api = new RippleAPI({ server: this._xrpServer })
+    this._secret = opts.secret
+    this._address = opts.address || deriveAddress(deriveKeypair(this._secret).publicKey)
+
+    this._peerAddress = opts.peerAddress // TODO: try to get this over the paychan?
+    this._bandwidth = opts.maxUnsecured || DEFAULT_BANDWIDTH
```
Comment (dappelt) [inline index.js]: `this._bandwidth` is not enforced. `_handleData` should check that the bandwidth is not exceeded before preparing a transfer.

File: index.js [hunk 2]
```diff
@@ -1,500 +1,411 @@
...
-  if (!chanId) {
-    ctx.plugin.debug('quering peer for incoming channel id')
-    try {
-      chanId = await ctx.rpc.call('ripple_channel_id', self.prefix, [])
-    } catch (err) { ctx.plugin.debug(err) }
 
-    if (!chanId) {
-      ctx.plugin.debug('cannot load incoming channel. Peer did not return incoming channel id')
-      return
+    if (!this._dataHandler) {
```
Comment (dappelt) [inline index.js]: Where is `_dataHandler` registered?

Comment (sharafian) [reply index.js]: plugin btp

File: index.js (hunk 1 above)
Comment (sharafian) [reply index.js]: That's now connector functionality, so it doesn't belong in this plugin at all. I've taken that line out.

File: store-wrapper.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  constructor (store) {
+    this._store = store
+    this._cache = new Map()
+    this._write = Promise.resolve()
+  }
+
+  async load (key) {
+    if (!this._store) return
+    if (this._cache.get(key)) return
+    this._cache.set(key, await this._store.get(key))
```
[the remark this answers is not in the document]
Comment (sharafian) [reply store-wrapper.js]: Fixed

**Pull request merged**